core/store/src/trie/outgoing_metadata.rs

```diff
@@ -0,0 +1,212 @@
+        let metadata = self.metadatas.get_mut(&shard_id).expect("Metadata should exist");
+        metadata.on_receipt_removed(receipt_size);
+    }
+}
+
+/// Keeps metadata about receipts stored in outgoing buffer to some shard.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OutgoingBufferMetadata {
+    pub receipt_group_sizes: ReceiptGroupSizes,
+}
+
+impl OutgoingBufferMetadata {
+    pub fn new(group_size_threshold: u32) -> Self {
+        OutgoingBufferMetadata {
+            receipt_group_sizes: ReceiptGroupSizes::V0(ReceiptGroupSizesV0::new(
+                group_size_threshold,
+            )),
+        }
+    }
+
+    /// Update the metadata when a new receipt is added at the end of the buffer.
+    pub fn on_receipt_buffered(&mut self, receipt_size: u64) {
+        match &mut self.receipt_group_sizes {
+            ReceiptGroupSizes::V0(v0) => v0.on_receipt_pushed(receipt_size),
+        }
+    }
+
+    /// Update the metadata when a receipt is removed from the front of the buffer.
+    pub fn on_receipt_removed(&mut self, receipt_size: u64) {
+        match &mut self.receipt_group_sizes {
+            ReceiptGroupSizes::V0(v0) => v0.on_receipt_popped(receipt_size),
+        }
+    }
+
+    /// Iterate over the sizes of receipts stored in the outgoing buffer.
+    /// Multiple consecutive receipts are grouped into a single group.
+    /// The iterator returns the size of each group, not individual receipts.
+    pub fn iter_receipt_group_sizes(&self) -> impl Iterator<Item = u64> + '_ {
+        match &self.receipt_group_sizes {
+            ReceiptGroupSizes::V0(v0) => v0.iter_receipt_group_sizes(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub enum ReceiptGroupSizes {
+    V0(ReceiptGroupSizesV0),
+}
+
+/// Keeps information about the size of receipts stored in the outgoing buffer.
+/// Consecutive receipts are grouped into groups and it keeps aggregate information about the groups,
+/// not individual receipts. Receipts are grouped to keep the size of the struct small.
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ReceiptGroupSizesV0 {
+    /// Receipts stored in the outgoing buffer, grouped into groups of consecutive receipts.
+    /// Each group stores aggregate information about the receipts in the group.
+    pub groups: VecDeque<ReceiptGroup>,
+    /// A new receipt group is started when the size of the last group exceeds this threshold.
+    /// All groups (except for the first and last one) have at least this size.
+    pub group_size_threshold: u32,
+}
+
+impl ReceiptGroupSizesV0 {
+    pub fn new(group_size_threshold: u32) -> Self {
+        Self { groups: VecDeque::new(), group_size_threshold }
+    }
+
+    /// Update the groups when a new receipt is added at the end of the buffer.
+    pub fn on_receipt_pushed(&mut self, receipt_size: u64) {
```

Comment:
Is this using the receipt size that is stored in the StateStoredReceipt? It's important for protocol upgrades if we ever decide to change the size accounting logic.

Reply:
Yes, it's the same size as in congestion control